src/AzureRenderManager/Microsoft.Identity.Web/StartupHelpers.cs

```diff
@@ -0,0 +1,228 @@
+                        context.Properties.Parameters.Remove(OpenIdConnectParameterNames.DomainHint);
+                    }
+
+                    // Additional claims
+                    if (context.Properties.Items.ContainsKey(OidcConstants.AdditionalClaims))
+                    {
+                        context.ProtocolMessage.SetParameter(OidcConstants.AdditionalClaims,
+                                                             context.Properties.Items[OidcConstants.AdditionalClaims]);
+                    }
+
+                    return Task.FromResult(0);
+                };
+            });
+            return services;
+        }
+
+        // Because we don't have a persistent token cache, after a web app
+        // restart the user can have a valid token set in a cookie, but the 
+        // server side TokenCache doesn't have the refresh token and hence we cannot
+        // get Graph or ARM resource tokens.
+        // For this reason we set a marker to indicate the user has logged into 
+        // this instance of the web app at least once.
+        private static Task SetIncarnationCacheKey(CookieSigningInContext context)
+        {
+            var cache = context.HttpContext.RequestServices.GetRequiredService<IMemoryCache>();
+            var cacheKey = CacheKey(GetObjectId(context.Principal.Claims));
+            if (!cache.TryGetValue(cacheKey, out var value))
+            {
+                // Set a user specific key to indicate they have logged in during this
+                // web app lifetime.
+                var incarnationId = Guid.NewGuid().ToString();
+                cache.Set<string>(cacheKey, incarnationId);
+            }
+            return Task.CompletedTask;
+        }
+
+        private static Task ValidateTokenAndCache(CookieValidatePrincipalContext context)
+        {
+            // If this user doesn't have a key in the cache then
+            // they don't have a refresh token in the TokenCache so we
+            // need to force a login by rejecting the principal.
+            var cache = context.HttpContext.RequestServices.GetRequiredService<IMemoryCache>();
+            if (!cache.TryGetValue(CacheKey(GetObjectId(context.Principal.Claims)), out string value))
+            {
+                context.RejectPrincipal();
+            }
+
+            if (context.Properties.Items.TryGetValue(".expires", out var expiryString)
+                && DateTimeOffset.TryParse(expiryString, out var expiresAt)
+                && expiresAt < DateTimeOffset.UtcNow)
+            {
+                context.RejectPrincipal();
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private static string GetObjectId(IEnumerable<Claim> claims)
+        {
+            return claims.FirstOrDefault(
+                c => c.Type == "http://schemas.microsoft.com/identity/claims/objectidentifier")?.Value;
```

Comment:
its also possibly that this is just "oid" in a JWT (lots of the code does conversion back and forth between short and long). However, if testing is showing it as ok its probably fine. I'd check what TokenAcquisition does for it key etc

Use `ClaimConstants.GetObjectId()` instead? or `ClaimsPrincipalExtension.GetMsalId()` instead?

Reply:
I've actually switched to using the MSAL account identifier (oid/tenant) and falling back to login hint.